### PR TITLE
[Record] SP8192 + SDClip + 3-Layer Depth Recurrence + EMA 0.9965 — val_bpb 1.0866

### PR DESCRIPTION
**Pull request by X-Abhishek-X**
## Record: SP8192 + SDClip + 3-Layer Depth Recurrence + EMA 0.9965 — val_bpb 1.0866

**val_bpb: 1.0866** (3-seed mean, std 0.0007) | **~15.98 MB** | 8×H100 SXM, 590s

### 3-Seed Results (8×H100 80GB SXM)

| Seed | Pre-quant BPB | **Sliding BPB (s64)** | Pruning | Artifact |
|------|---------------|----------------------|---------|----------|
| 42 | 1.0874 | **1.0873** | None | 15,981,300 B |
| 1337 | 1.0865 | **1.0866** | None | 15,978,870 B |
| 2024 | — | **1.0859** | None | — |
| **Mean** | | **1.0866 (std 0.0007)** | **Zero** | |

Current merged SOTA: **1.1147** (PR #1019). Delta: **−0.0281 BPB**.

### Key Changes (over PR #1445, this author)

| Change | PR #1445 | This | Impact |
|--------|----------|------|--------|
| **Tokenizer** | SP4096 | **SP8192** | Larger vocab, better context |
| **Quantization** | Percentile search | **SDClip (c = k·std)** | Zero pruning, better rate-distortion |

### SDClip Quantization

Replaces multi-percentile clip search with `clip = k · std(row)` (PR #1394):
- **k=12.85** for int6 matrices, **k=20.0** for int8 embeddings
- Directly accounts for compressed size, not just reconstruction error
- One GPTQ pass per matrix instead of 5
- Result: **zero selective pruning** — model fits natively under 16MB

### Full Stack

| Parameter | Value | Source |
|-----------|-------|--------|
| Tokenizer | SP8192 | This work |
| SDClip k (matrices/embed) | 12.85 / 20.0 | PR #1394, this work |
| Recurrence layers | 3,4,5 (14 virtual) | PR #1331 |
| Weight decay | 0.095 | PR #1331 |
| Matrix LR | 0.022 | PR #1331 |
| EMA decay | 0.9965 | PR #1421 (this author) |
| Recurrence start | step 2000 | PR #1445 (this author) |
| Warmdown fraction | 0.72 | PR #1445 (this author) |

### Architecture

- 11L, 512-dim, 8 heads (4 KV), depth recurrence (3,4,5), 14 virtual layers
- Skip gates, parallel residuals from layer 7, QK-Gain 5.0
- XSA all 11 layers, LeakyReLU(0.5)², VE128 (layers 9,10)
- Tied embeddings, logit softcap=30.0

### Training

- FlashAttention 3, Muon (lr=0.022, WD=0.095), Adam/AdamW (fused=True)
- Warmdown: 72%, EMA=0.9965, Wallclock: 590s

### Quantization

- Full Hessian GPTQ + Cholesky + actorder
- SDClip (c = k·std) — int6 matrices, int8 embeddings
- Brotli compression, zero selective pruning

### Run Command

```bash
SEED=42 VOCAB_SIZE=8192 \
DATA_PATH=./data/datasets/fineweb10B_sp8192/ \
TOKENIZER_PATH=./data/tokenizers/fineweb_8192_bpe.model \
RECUR_START_STEP=2000 WARMDOWN_FRAC=0.72 \
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### Credits

- **SDClip + SP8192 baseline**: PR #1394 by @clarkkev
- **Base architecture**: PR #1334 by @aryanbhosale
- **3-layer recurrence + WD/LR**: PR #1331
- **EMA 0.9965 + early recurrence + warmdown**: PRs #1421, #1445 by @X-Abhishek-X (this author)

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — [Record] SP8192 + SDClip + 3-Layer Depth Recurrence + EMA 0.9965 — val_bpb 1.0866

**BPB:** 1.0866 | **Compliance:** LOOKS CLEAN — score-first-per-chunk TTT (legal #1416/#1423 pattern)

**What I found in the code** (head SHA `e7405f8e77f1`, file `records/track_10min_16mb/2026-04-08_SP8192_SDClip_3LayerRecur_EMA0.9965_1.0866/train_gpt.py`):

The TTT path at line **1508** implements the score-first-per-chunk pattern: each chunk is scored under `torch.no_grad()` / `inference_mode()` before the `base_model.train()` + SGD adaptation runs on that same chunk, with an `is_last_chunk` guard so the final chunk gets no adaptation pass. This is the structural shape the legal frontier uses (PRs #1416 erichroepke, #1423 aryanbhosale).

Per **Issue #402 and Issue #677**, TTT is legal when each token is scored before the adapter updates on it, and that's what the code does here — chunk `ci` is scored under weights adapted only on chunks `0..ci-1`. No `prequant_ttt_adapt_adamw(val_tokens, ...)` multi-epoch fine-tune, no scored-region SLOT, no target-in-key n-gram cache.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 10.78s, dim=512, layers=11, vocab=4096, code=83119 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending standard checks (3-seed validation, 16MB artifact cap, 10-min wallclock on 8×H100 SXM). The compliance picture matches the legal reference frontier and no flags were raised by the classification pass.

*Auto-classification caveat:* this review was drafted by the AST-based classifier against a template derived from manually-reviewed cluster PRs (#1420, #1450, #1487, #1541, #1529, #1533, #1518). If I've misread a subtlety in your eval path — e.g., multi-epoch TTT that I mistook for single-pass, or a target-in-key lookup I missed in a helper function — please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 10.78s, dim=512, layers=11, vocab=4096, code=83119 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*
